lib/AST/Module.cpp

```diff
@@ -353,17 +375,32 @@ void SourceLookupCache::populateAuxiliaryDeclCache() {
 
       for (const auto *macro : found->second) {
         macro->getIntroducedNames(MacroRole::Peer,
-                                  decl, introducedNames[attr]);
+                                  dyn_cast<ValueDecl>(decl),
+                                  introducedNames[attr]);
+      }
+    }
+
+    if (auto *med = dyn_cast<MacroExpansionDecl>(decl)) {
+      UnresolvedMacroReference macroRef(med);
+      auto macroName = macroRef.getMacroName().getBaseIdentifier();
+
+      auto found = MacroDecls.find(macroName);
+      if (found == MacroDecls.end())
+        continue;
+
+      for (const auto *macro : found->second) {
+        macro->getIntroducedNames(MacroRole::Declaration,
+                                  /*attachedTo*/ nullptr,
```

Comment:
Nit: the `dyn_cast<ValueDecl>(decl)` above is already tolerant to `MacroExpansionDecl`, because the dynamic case will just return `nullptr`. I had intended for the code above to instead conditionalize the macro role and macro reference, instead of the entire `if` body